### PR TITLE
Merge ISSampler to garage.sampler

### PR DESCRIPTION
**Pull request by zequnyu**
- relocated contrib.alexbeloi ISSampler and examples
- added test to ISSampler
- updated code and doc styles

Address #685.

### CODE REVIEW

File: src/garage/sampler/is_sampler.py [hunk 1]
```diff
@@ -84,54 +85,54 @@ def obtain_samples(self, itr):
         if self._is_itr and not self.skip_is_itrs:
             paths = self.obtain_is_samples(itr)
         else:
-            paths = super(ISSampler, self).obtain_samples(itr)
+            paths = super(ISSampler, self).obtain_samples(
```
Comment (ryanjulian) [inline src/garage/sampler/is_sampler.py]: i think this can just be `super().obtain_samples(...)`

File: src/garage/sampler/is_sampler.py [hunk 2]
```diff
@@ -84,54 +85,54 @@ def obtain_samples(self, itr):
...
+                itr, batch_size, whole_paths)
             if not self.skip_is_itrs:
                 self.add_history(self.algo.policy.distribution, paths)
 
         self._is_itr = (self._is_itr + 1) % 2
         return paths
 
-    def obtain_is_samples(self, itr):
+    def obtain_is_samples(self, itr, batch_size=None, whole_paths=True):
```
Comment (ryanjulian) [inline src/garage/sampler/is_sampler.py]: maybe make this _private?

File: src/garage/sampler/is_sampler.py [hunk 3]
```diff
@@ -172,11 +173,7 @@ def sample_isweighted_paths(
             path['rewards'] *= is_ratio
 
         if ess_threshold:
-            if kong_ess(is_weights) < ess_threshold:
+            if len(is_weights) / (1 + var(is_weights)) < ess_threshold:
```
Comment (ryanjulian) [inline src/garage/sampler/is_sampler.py]: if you want to eliminate the helper function, please leave a comment that this is the effective sample size estimate used by Kong 1992:

```
Kong, Augustine. "A note on importance sampling using standardized weights." University of Chicago, Dept. of Statistics, Tech. Rep 348 (1992).
```

File: src/garage/sampler/is_sampler.py [hunk 4]
```diff
@@ -84,54 +85,54 @@ def obtain_samples(self, itr):
...
-            hist_variance_penalty=0.0,
-            max_is_ratio=10,
-            ess_threshold=0,
-    ):
+        if len(paths) > batch_size:
+            paths = random.sample(paths, batch_size)
+
+        return paths if whole_paths else truncate_paths(paths, batch_size)
+
+    def sample_isweighted_paths(self,
```
Comment (ryanjulian) [inline src/garage/sampler/is_sampler.py]: probably make this _private

File: src/garage/sampler/batch_sampler.py [hunk 5]
```diff
@@ -8,32 +8,32 @@ class BatchSampler(BaseSampler):
...
 
     def shutdown_worker(self):
         """Shutdown worker function."""
         parallel_sampler.terminate_task(scope=self.algo.scope)
 
-    def obtain_samples(self, itr):
+    def obtain_samples(self, itr, batch_size=None, whole_paths=True):
         """Obtain samples function."""
+        if not batch_size:
+            batch_size = self.algo.max_path_length
```
Comment (ryanjulian) [inline src/garage/sampler/batch_sampler.py]: does this field exist on algo anymore?

Comment (zequnyu) [reply src/garage/sampler/batch_sampler.py]: For `max_path_length` yes.

File: src/garage/sampler/is_sampler.py [hunk 6]
```diff
@@ -172,11 +172,11 @@ def sample_isweighted_paths(
             path['rewards'] *= is_ratio
```
Comment (zhanpenghe) [inline src/garage/sampler/is_sampler.py]: Is this used at all?

Comment (zequnyu) [reply src/garage/sampler/is_sampler.py]: Depending on algorithms it use I think yes.

Comment (zhanpenghe) [reply src/garage/sampler/is_sampler.py]: Oh yes it's passed by ref

**Comment by zequnyu** [comment]
Please take a look at the new change: reset the global `singleton_pool` for tf test cases.

File: tests/fixtures/fixtures.py [hunk 7]
```diff
@@ -29,6 +29,10 @@ def setUp(self):
         logger.add_output(NullOutput())
         deterministic.set_seed(1)
 
+        # initialize global singleton_pool for each test case
+        from garage.sampler import singleton_pool
+        singleton_pool.initialize(1)
```
Comment (naeioi) [inline tests/fixtures/fixtures.py]: Does this help to fix any hanging bug?

Comment (zequnyu) [reply tests/fixtures/fixtures.py]: This [test](https://github.com/rlworkgroup/garage/issues/713#issuecomment-498942326) would not pass without it. The `test_batch_sampler()` is logger's problem. I'll take a deeper look to see if we can do something to resolve it. Essentially this helps with any hanging bugs due to the global `singleton_pool`, similar to the `snapshotter` one.

**Pull request merged**